### PR TITLE
TSLint → ESLint

### LINKED ISSUE #16053: Consider a switch from tslint to eslint
**What would you like to be added**:
I would like us to consider a switch from from tslint to eslint

**Why is this needed**:
Given the recent posts from tslint (https://github.com/Microsoft/TypeScript/issues/29288) and eslint (https://eslint.org/blog/2019/01/future-typescript-eslint) teams it might be time to switch?

As an additional perk we can consider to integrate prettier as an eslint rule as well (https://prettier.io/docs/en/eslint.html), so we could use `eslint --fix` for all the things

### LINKED ISSUE #20430: grafana/toolkit : Linting *.ts files make them blank
Hi Domink @dprokop,

Some time, with a toolkit:dev, the ts file is emptied, since i save all ts file before to build.
 
    "@grafana/data": "^6.6.0-pre-e0a2d4beac",
    "@grafana/runtime": "^6.6.0-pre-e0a2d4beac",
    "@grafana/toolkit": "^6.6.0-pre-e0a2d4beac",

An example here, ts is empty after : 
- Linting

  ERROR: /rulesHandler.ts[41:23]: != should be !==
  ERROR: /rulesHandler.ts[17:0]: jsdoc is not formatted correctly on this line
  ERROR: /rulesHandler.ts[103:0]: jsdoc is not formatted correctly on this line
  ERROR: /rulesHandler.ts[155:0]: jsdoc is not formatted correctly on this line

  Trace: Error: 4 linting errors found in 1 files
      at D:\Dev\workspace\grafana\plugin\grafana-flowcharting\node_modules\@grafana\toolkit\src\cli\tasks\plugin.build.js:172:23
      at step (D:\Dev\workspace\grafana\plugin\grafana-flowcharting\node_modules\tslib\tslib.js:136:27)
      at Object.next (D:\Dev\workspace\grafana\plugin\grafana-flowcharting\node_modules\tslib\tslib.js:117:57)
      at D:\Dev\workspace\grafana\plugin\grafana-flowcharting\node_modules\tslib\tslib.js:110:75
      at new Promise (<anonymous>)
      at Object.__awaiter (D:\Dev\workspace\grafana\plugin\grafana-flowcharting\node_modules\tslib\tslib.js:106:16)
      at D:\Dev\workspace\grafana\plugin\grafana-flowcharting\node_modules\@grafana\toolkit\src\cli\tasks\plugin.build.js:139:20
      at D:\Dev\workspace\grafana\plugin\grafana-flowcharting\node_modules\@grafana\toolkit\src\cli\utils\useSpinner.js:17:42
      at step (D:\Dev\workspace\grafana\plugin\grafana-flowcharting\node_modules\tslib\tslib.js:136:27)
      at Object.next (D:\Dev\workspace\grafana\plugin\grafana-flowcharting\node_modules\tslib\tslib.js:117:57)
      at D:\Dev\workspace\grafana\plugin\grafana-flowcharting\node_modules\tslib\tslib.js:110:75
      at new Promise (<anonymous>)
      at Object.__awaiter (D:\Dev\workspace\grafana\plugin\grafana-flowcharting\node_modules\tslib\tslib.js:106:16)
      at Object.lintPlugin (D:\Dev\workspace\grafana\plugin\grafana-flowcharting\node_modules\@grafana\toolkit\src\cli\utils\useSpinner.js:7:48)
      at D:\Dev\workspace\grafana\plugin\grafana-flowcharting\node_modules\@grafana\toolkit\src\cli\tasks\plugin.dev.js:76:53
      at step (D:\Dev\workspace\grafana\plugin\grafana-flowcharting\node_modules\tslib\tslib.js:136:27)
      at Console.trace (internal/console/constructor.js:336:11)
      at D:\Dev\workspace\grafana\plugin\grafana-flowcharting\node_modules\@grafana\toolkit\src\cli\utils\useSpinner.js:24:29
      at step (D:\Dev\workspace\grafana\plugin\grafana-flowcharting\node_modules\tslib\tslib.js:136:27)
      at Object.throw (D:\Dev\workspace\grafana\plugin\grafana-flowcharting\node_modules\tslib\tslib.js:117:57)
      at rejected (D:\Dev\workspace\grafana\plugin\grafana-flowcharting\node_modules\tslib\tslib.js:108:69)
× 4 linting errors found in 1 files
error Command failed with exit code 1.

### PR DESCRIPTION
**Pull request by stevenvachon**
**What this PR does / why we need it**:
Migrates from the now deprecated [tslint](https://npmjs.com/tslint) to [eslint](https://npmjs.com/eslint) (and the necessary plugins for TypeScript).  Some basic formatting rules were moved out of "lint" configs and into `.editorconfig` for reduced redundancy.

With that, some supportive changes were necessary (#21441, #21586) ~~and they should be merged before this~~.

**Which issue(s) this PR fixes**:
Fixes #16053, #19345
Indirectly fixes #20430

**Special notes for your reviewer**:
One liberty taken was to separate our [eslintrc](https://eslint.org/docs/user-guide/configuring) and [tsconfig](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html) to their own packages: [@grafana/eslint-config](https://npmjs.com/@grafana/eslint-config) and [@grafana/tsconfig](https://npmjs.com/@grafana/tsconfig). Instead of parent referencing (`../`) directories in many places, a package name is used and fewer files overall are necessary. For me personally, this simplified my mental graph of the codebase that I am still learning, which I think will translate to colleagues and third-party plugin developers.

### CODE REVIEW

**Comment by ryantxu** [comment]
> @torkelo @dprokop -- I was told that you might be concerned with this change.

😄 I said they will have opinions!

My biggest concern is with polluting the package names with many small packages.  That may be a non-concern or something we can handle with folders or prefixes.  This currently proposes:
image

Maybe changing the folder names to: `config-tslint` and `config-tsconfig`?  would group them better?

**Comment by torkelo** [comment]
Cant the tsconfig and eslint config live in toolkit package or do they have to be in their own packages?

**Comment by stevenvachon** [comment]
I don't think that it makes sense to depend on the toolkit package for lints to work correctly within other unrelated packages. The purpose of a monorepo is to easily make changes across multiple—but still isolated—packages, not to treat them as a single project.

**Comment by torkelo** [comment]
But the toolkit package is designed to contain some underlying building blocks and build scripts, shared config etc, exactly things tsconfig and lint config so creating a separate packages with a single file seems too granular, but maybe warranted / good idea, just will increase time running workspace commands

**Comment by dprokop** [comment]
> But the toolkit package is designed to contain some underlying building blocks and build scripts, shared config etc, exactly things tsconfig and lint config so creating a separate packages with a single file seems too granular, but maybe warranted / good idea, just will increase time running workspace commands

@torkelo as far as I see this PR does not introduce changes to the toolkit's internal tsconfig (common config for plugin developers) but only for how the packages are using the shared config. So grafana-toolkit still uses tslint as a driver for plugin linting. But I don't think changing this is a goal of this PR, isn't it @stevenvachon? I think having the tslint/tsconfig packages would be useful for plugin devs so that they don't have to extend tsconfig in a way they need to now:

```
  "extends": "./node_modules/@grafana/toolkit/src/config/tsconfig.plugin.json",
```
would become
```
  "extends": "@grafana/tsconfig",
```

I think the shared packages are not necessarily a bad idea- it's a common practice to have configs available as packages. But those packages will have to be also published to npm for the plugin developers to take advantage from. 

@stevenvachon do I understand correctly that this PR replaces TSLint with ESLint in the core Grafana too?

File: package.json [hunk 1]
```diff
@@ -157,7 +162,7 @@
     "watch": "yarn start -d watch,start core:start --watchTheme ",
     "build": "grunt build",
     "test": "grunt test",
-    "tslint": "tslint -c tslint.json --project tsconfig.json",
+    "tslint": "eslint public/app public/e2e-test public/test public/vendor --ext=.js,.ts,.tsx",
```
Comment (torkelo) [inline package.json]: don't think we need to lint public/vendor, old vendored libs that will have hundreds of errors

File: packages/grafana-eslint-config/test.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+'use strict';
+const {expect} = require('chai');
```
Comment (torkelo) [inline packages/grafana-eslint-config/test.js]: This looks like prettier is not formatting, need to update the root package.json prettier commands to also run on js files :)

**Review by torkelo — COMMENTED** [review]
yea, the extends become a bit shorter if they are at the root of their own package. 

The CI build error looks strange, yarn seems not to be able to link in the package

**Comment by stevenvachon** [comment]
> So grafana-toolkit still uses tslint as a driver for plugin linting.

No, grafana-toolkit will soon use ESLint as well. TSLint is deprecated.

> @stevenvachon do I understand correctly that this PR replaces TSLint with ESLint in the core Grafana too?

Correct.

File: packages/grafana-eslint-config/test.js (hunk 2 above)
Comment (stevenvachon) [reply packages/grafana-eslint-config/test.js]: Yeah, that's just something temporary that I'd created quickly to outline my plans.

**Comment by torkelo** [comment]
once this is merged, would be nice to get it to output junit report so circle ci can report eslint error in a formatted way at the top: 
https://www.viget.com/articles/using-junit-on-circleci-2-0-with-jest-and-eslint/ 

i recently did this for frontend tests, and it greatly improves finding failures in CI results:
https://github.com/grafana/grafana/pull/21084

**Comment by ryantxu** [comment]
Looking at this a bit more...  I don't like that we would be adding two packages `grafana-eslint-config` and `grafana-tsconfig` to the main grafana build process.

It also seems a little weird to have the config files version locked to the grafana version.

Maybe a better option is to create two new repos:
```
eslint-config-grafana
tsconfig-grafana
```
And publish to npm with that name (not @grafana/...)  This would follow what airbnb does

@dprokop thoughts?

**Comment by torkelo** [comment]
Maybe these two packages can be seperate npm packages @grafana/es-lint-config & @grafana/tsconfig but maybe we can put them in a separate repo (grafana/npm-packages)? we wont be able to reuse our npm publish scripts but that could be minimal?

But then again, how much will adding these two projects add to the build process? They wont require any building really, only publish? 

@dprokop what do you think?

**Comment by dprokop** [comment]
Regarding eslint config. There is an official ESLint recommendation about the naming that can be found here: https://eslint.org/docs/developer-guide/shareable-configs#creating-a-shareable-config

**Naming**
In summary the package name should be prefixed with `eslint-config-` which in our case the name shoud be either `eslint-config-grafana` or `@grafana/eslint-config-grafana`. For the matter of consistency and organization scoping I would go with `@grafana/eslint-config-grafana`, and the same with tsconfig configuration: `@grafana/tsconfig-grafana`. 

**Publishing**
This should require minor changes so that eslint/tsconfig packages are published only when the actual package changes. Currently ALL of the @grafana/* packages are published (canary) when there are changes detected in ANY of the packages. I guess the config should be released independently of that rule. Also, there will be no build step for these two packages so publishing them in the main build pipeline should not impact us much - especially taking the config changes rate into account.

**Versioning**
Not sure what's the best approach here. Agree with @ryantxu that having the config version locked to Grafana version feels weird. Some thoughts on this @torkelo @stevenvachon ?

**Comment by ryantxu** [comment]
+1 for:
```
@grafana/eslint-config-grafana
@grafana/tsconfig-grafana
```

For versioning, I would put each in its own repository, I suggest matching the names above.

**Comment by torkelo** [comment]
I think they can be in this repo,I talked to dominik and we do not need to publish these packages on every build , so it will not slow down the build.

And having to duplicate our npm package publish scripts and pipeline seems bad

**Comment by ryantxu** [comment]
My argument for having them in a different repo is that the deployment and versions are totally different -- I think it should be explicit and manual in a way the packages/* are not.

If we do keep it in the same repo, maybe a different folder?

**Comment by dprokop** [comment]
I think if we are deciding to keep these two in the same repo we should keep them in the same directory as other packages. These packages will not be automatically released with our pipeline unless directly specified in the release scripts, so we should be fine here. The one thing that might be problematic is the `packages:prepare` task that updates the cross-package versions to the lerna.json specified one, which means that config package versions will be updated as well - not sure what is the workaround for that apart from manually modyfing the versions back to the correct one in the script.

Having these two packages in another directory is not so good as we will introduce inconsistency in package.json and lerna.json about which packages(versioning) are managed by what pipeline (core packages versioning managed by lerna, config packages managed manually). Also we would need to have two directories for workspaces which feels weird.

@torkelo we will have to have separate publishing scripts for config packages anyways as core packages are managed by Lerna (for config packages it's actually not a script, but just an npm script running `npm publish`).

Taking into account the frequency of changes in the config packages, different versioning schema and simplified publishing I think we can have them as separate repo(s).

File: packages/grafana-toolkit/src/cli/tasks/plugin.build.ts [hunk 3]
```diff
@@ -1,173 +1,97 @@
...
-// @ts-ignore
 const typecheckPlugin = useSpinner<void>('Typechecking', async () => {
   await execa('tsc', ['--noEmit']);
 });
 
-const getTypescriptSources = () => {
-  const globPattern = path.resolve(process.cwd(), 'src/**/*.+(ts|tsx)');
-  return glob.sync(globPattern);
-};
+const getTypescriptSources = () => globby(resolvePath(process.cwd(), 'src/**/*.+(ts|tsx)'));
```
Comment (ryantxu) [inline packages/grafana-toolkit/src/cli/tasks/plugin.build.ts]: are these `globby` changes related to TSLint -> ESLint?  

I like the change, but maybe we can pull that into a different PR and get it committed quickly?

Comment (stevenvachon) [reply packages/grafana-toolkit/src/cli/tasks/plugin.build.ts]: They'll be split out to a separate "cleanup" PR when all the work is done.  It's currently all part of a related train of thought.

**Comment by torkelo** [comment]
> @torkelo we will have to have separate publishing scripts for config packages anyways as core packages are managed by Lerna (for config packages it's actually not a script, but just an npm script running npm publish). 

That settles it then, seperate repo!

**Comment by ryantxu** [comment]
I just created:
https://github.com/grafana/eslint-config-grafana
https://github.com/grafana/tsconfig-grafana

and copied the files from this PR to each of them.  @stevenvachon -- can you update the structure as needed.  (not sure when it is easiest to do... and to start publishing to npm)

File: packages/grafana-data/package.json [hunk 4]
```diff
@@ -14,13 +14,15 @@
...
-    "tslint": "tslint -c tslint.json --project tsconfig.json",
+    "lint": "eslint src/ --ext=.js,.ts,.tsx",
     "typecheck": "tsc --noEmit",
     "clean": "rimraf ./dist ./compiled",
     "bundle": "rollup -c rollup.config.ts",
     "build": "grafana-toolkit package:build --scope=data"
   },
   "devDependencies": {
+    "@grafana/eslint-config": "^1.0.0-pre",
+    "@grafana/tsconfig": "^1.0.0-pre",
```
Comment (ryantxu) [inline packages/grafana-data/package.json]: Should this be: [@grafana/tsconfig-grafana](https://github.com/grafana/tsconfig-grafana/blob/master/package.json#L4)

File: packages/grafana-data/package.json [hunk 5]
```diff
@@ -14,13 +14,15 @@
...
   "scripts": {
-    "tslint": "tslint -c tslint.json --project tsconfig.json",
+    "lint": "eslint src/ --ext=.js,.ts,.tsx",
     "typecheck": "tsc --noEmit",
     "clean": "rimraf ./dist ./compiled",
     "bundle": "rollup -c rollup.config.ts",
     "build": "grafana-toolkit package:build --scope=data"
   },
   "devDependencies": {
+    "@grafana/eslint-config": "^1.0.0-pre",
```
Comment (ryantxu) [inline packages/grafana-data/package.json]: Should this be [@grafana/eslint-config-grafana](https://github.com/grafana/eslint-config-grafana/blob/master/package.json#L4)

File: packages/grafana-data/package.json (hunk 4 above)
Comment (stevenvachon) [reply packages/grafana-data/package.json]: Yes, I haven't renamed/moved them yet.

File: packages/grafana-data/package.json (hunk 5 above)
Comment (stevenvachon) [reply packages/grafana-data/package.json]: Yes, I haven't renamed/moved them yet.

**Comment by stevenvachon** [comment]
Hmm:
> npm scoped modules are also supported, by naming or prefixing the module with `@scope/eslint-config`, such as `@scope/eslint-config` **or** `@scope/eslint-config-myconfig`.

I think `@scope/eslint-config` is less redundant.

**Comment by torkelo** [comment]
yea, @scope/eslint-config looks fine

**Comment by dprokop** [comment]
@stevenvachon +1 for @scope/eslint-config

**Comment by stevenvachon** [comment]
The following scripts are working and passing:
* `yarn build`
* `yarn lint`
* `yarn typecheck`
* `yarn packages:build`
* `yarn packages:lint`
* `yarn packages:typecheck`

The following fails miserably:
* `yarn test`

If anyone (@dprokop, @ryantxu, @torkelo) can take a look to help solve this, would likely expedite completion.

**Comment by ryantxu** [comment]
for the "StringMap expected string key" maybe:
https://medium.com/@gavin.blog/how-to-solve-ng-annotate-stringmap-expected-string-key-error-f0711c7c84db

File: package.json [hunk 6]
```diff
@@ -59,6 +59,8 @@
     "@types/slate-react": "0.22.5",
     "@types/testing-library__react-hooks": "^3.1.0",
     "@types/tinycolor2": "1.4.2",
+    "@typescript-eslint/eslint-plugin": "^2.18.0",
```
Comment (dprokop) [inline package.json]:
```suggestion
    "@typescript-eslint/eslint-plugin": "2.18.0",
```

File: package.json [hunk 7]
```diff
@@ -59,6 +59,8 @@
...
+    "@typescript-eslint/parser": "^2.18.0",
```
Comment (dprokop) [inline package.json]:
```suggestion
    "@typescript-eslint/parser": "2.18.0",
```

File: package.json [hunk 8]
```diff
@@ -73,6 +75,12 @@
     "enzyme-to-json": "3.3.5",
     "es6-promise": "3.3.1",
     "es6-shim": "0.35.5",
+    "eslint": "^6.8.0",
```
Comment (dprokop) [inline package.json]: Please fix the version of dependencies

File: public/app/core/services/ng_react.ts [hunk 9]
```diff
@@ -54,7 +54,7 @@ function applied(fn: any, scope: any) {
   if (fn.wrappedInApply) {
     return fn;
   }
-  //tslint:disable-next-line:only-arrow-functions
+  // this had the equivalent of `eslint-disable-next-line prefer-arrow/prefer-arrow-functions`
```
Comment (dprokop) [inline public/app/core/services/ng_react.ts]: Not sure what you mean by this comment?

**Review by dprokop — COMMENTED** [review]
This is looking good, great stuff @stevenvachon . There is one thing that I find missing. In the packages README I'd like to see a guide on how and where from to publish the new configs. This should also be documented in config's READMEs. 

This also introduces a *breaking change* in the grafana/toolkit. I think we need to add some kind of guide on how to migrate. Maybe try automating the tslint/tsconfig -> eslint migration in toolkit lint task?

File: package.json (hunk 6 above)
Comment (stevenvachon) [reply package.json]: Why when we have a lockfile?

File: public/app/core/services/ng_react.ts (hunk 9 above)
Comment (stevenvachon) [reply public/app/core/services/ng_react.ts]: Some tslint rules are slightly different in eslint (and typescript-eslint), which would've resulted in many source code file changes. They'll be handled soon in another PR, but I wanted to keep a record of the previous state without retaining any tslint comments.

Comment (dprokop) [reply public/app/core/services/ng_react.ts]: gottcha

File: package.json (hunk 6 above)
Comment (dprokop) [reply package.json]: We had couple of issues in the past when regenerated yarn.lock(from scratch) was referencing versions which were causing troubles.

**Review by ryantxu — APPROVED** [review]
LGTM -- lets fix any migration guides in a follow on issue
